### PR TITLE
WIP: Python API to PROCESS

### LINKED ISSUE #151: Use Python interface to PROCESS
In GitLab by @DanShort12 on Jan 8, 2021, 16:55

The PROCESS team have recently released their Python interface. We should look to leverage that in order to simplify our interaction with PROCESS. A few points to consider:

* Should we keep the ability to run PROCESS directly (in the longer term - it will likely be necessary to support running in both modes while we implement the new interface)?
* Are there any significant implications for distributing PROCESS within bluemira i.e. can we remove the PROCESS source code?
* We will likely still use a template IN.DAT for a given family of reactors (and just override what we need to for a given run).
* Do we want to be able to control optimization variables and constraints via bluemira or will that be done via the template IN.DAT?

### LINKED ISSUE #1685: Update default PROCESS install to more recent version
## Description of issue / requirement to address

We should keep up with the PROCESS changes. A lot has changed recently with PROCESS so it will probably need a little bit of work.

## Proposed solution

Ask PROCESS to cut a new release. Change our default installation to that version

## Alternative solutions

Leave it on v2.3

### PR DESCRIPTION
**Pull request by CoronelBuendia**
## Linked Issues

Closes #1685 
Closes #151 

## Checklist

I confirm that I have completed the following checks:

- [ ] Tests run locally and pass `pytest tests --reactor`
- [ ] Code quality checks run locally and pass `flake8` and `black .`
- [ ] Documentation built locally and checked `sphinx-build -W documentation/source documentation/build`

### CODE REVIEW

File: bluemira/codes/process/_equation_variable_mapping.py [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with bluemira; if not, see <https://www.gnu.org/licenses/>.
+"""
+Death to PROCESS integers
+"""
+
+OBJECTIVE_EQ_MAPPING = {
```
Comment (CoronelBuendia) [inline bluemira/codes/process/_equation_variable_mapping.py:25]: Could be an `Enum`...

File: bluemira/codes/process/_equation_variable_mapping.py [hunk 2]
```diff
@@ -0,0 +1,379 @@
...
+    # note: FoM should be minimised only!
+    "pnetel": 17,  # net electrical output
+    "NULL": 18,  # Null Figure of Merit
+    "Q_tpulse": 19,  # linear combination of big Q and pulse length (maximised)
+    # note: FoM should be minimised only!
+}
+
+OBJECTIVE_MIN_ONLY = [16, 19]
+
+CONSTRAINT_EQ_MAPPING = {
```
Comment (CoronelBuendia) [inline bluemira/codes/process/_equation_variable_mapping.py:50]: Could be an `Enum`...

File: bluemira/codes/process/template_builder.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+            )
+
+        return ProcessInputs(
+            bounds=self.bounds,
+            icc=self.icc,
+            ixc=self.ixc,
+            minmax=self.minmax,
+            ioptimz=self.ioptimiz,
+            epsvmc=self.epsvmc,
+            maxcal=self.maxcal,
```
Comment (CoronelBuendia) [inline bluemira/codes/process/template_builder.py:234]: forgot `fimp=self.fimp`... sorry

File: bluemira/codes/process/_equation_variable_mapping.py (hunk 1 above)
Comment (je-cook) [reply bluemira/codes/process/_equation_variable_mapping.py:25]: possibly using `MappingProxyType` for an immutable dictionary: https://adamj.eu/tech/2022/01/05/how-to-make-immutable-dict-in-python/ . Possibly overkill...

**Review by je-cook — CHANGES_REQUESTED** [review]
I think they're all typing or immutability changes. Overall happy though

File: bluemira/codes/process/_model_mapping.py [hunk 4]
```diff
@@ -0,0 +1,1117 @@
...
+
+    @classproperty
+    def switch_name(self) -> str:
+        """
+        PROCESS switch name
+        """
+        return "cost_model"
+
+    TETRA_1990 = 0
+    KOVARI_2015 = 1
```
[the remark this answers is not in the document]
Comment (CoronelBuendia) [reply bluemira/codes/process/_model_mapping.py:1101]: Yeah, this looks like a workable solution, and probably in the meantime PROCESS will evolve to be to accommodate custom models / constraint equations / objective functions dynamically somehow.

For now, we just need a Python API to what is actually there.

**Review by je-cook — APPROVED** [review]

**Comment by je-cook** [comment]
the test failures look like my fault, I'll go fix them

**Pull request merged**